### PR TITLE
update rl_reinforce.py

### PR DESCRIPTION
**Pull request by r08922196**
My reward design:
1. In the beginning, `delta_S` and `I` is more critical.
2. In the beginning, `I` should be given more penalty.
3. In the end, the remaining `S` would be given more reward. 

> For clarity, this kinda design is not for evaluation.
---
1. Update config.json to formulate more critical situation and let agent to learn.
2. Update reward function in class environment.

For training,
You could try the following training hyper-parameters:
```=json
{
    "iterations": 50,
    "num_rollouts": 20,
    "max_steps": 300,
    "learning_rate": 5e-3,
    "gamma": 0.99,
}
```
The results are as followed.
Iteration0

Iteration10

Iteration20

Iteration30

Iteration40

Iteration50
Iteration60
Iteration60
image

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
 {
     "beta": 0.015,
     "S": 23000000,
-    "E": 6,
+    "E": 100,
```
Comment (r08922a27) [inline config.json:4]: N_total 要記得更新。

File: config.json [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
     "E_move": 0,
     "Q": 0,
     "Q_move": 0,
-    "I": 6,
+    "I": 30,
```
Comment (r08922a27) [inline config.json:8]: N_total 要記得更新。

File: rl_reinforce.py [hunk 3]
```diff
@@ -302,24 +307,42 @@ def step(self, s, a, t):
...
+        delta_E_move = (SE_move - EI_move - EQ_move) / self._config['N_total']
+
+        delta_Q = (EQ - QI) / self._config['N_total']
+
+        delta_Q_move = (EQ_move - QI_move) / self._config['N_total']
+
+        delta_I = (EI + EI_move + QI + QI_move - IR) / self._config['N_total']
+
+        # the delta of each seir component is more important in the begining.
+        seir_delta_reward = (5000 / (t + 10)) * (3 * delta_S - 2 * delta_I  - (delta_Q + delta_Q_move)- (delta_E + delta_E_move))
```
Comment (r08922a27) [inline rl_reinforce.py:324]: t + 10 是為了因為 begining 嗎？(有點看不懂 t + 10 用意)

請問 3 跟 2 是調出來的權重還是有什麼其他想法呢？

File: rl_reinforce.py [hunk 4]
```diff
@@ -302,24 +307,42 @@ def step(self, s, a, t):
...
         
-        if s[N_QUARANTINE] == self._config['MAX_Q']:
-           
-           reward -= 0.01
+        # the state of S is more important in the end of period.
+        seir_reward = math.log((t + 31) / 5) * (3 * self.S)
```
Comment (r08922a27) [inline rl_reinforce.py:327]: t + 31 的 31 是指什麼？跟 end 的關係是什麼？

File: rl_reinforce.py [hunk 5]
```diff
@@ -302,24 +307,42 @@ def step(self, s, a, t):
...
+        
+        # the state of other components in seir model is more important in the begining.
+        seir_reward = (100 / (t + 10)) * (- 2 * self.I - (self.Q + self.Q_move) - (self.E + self.E_move))
+
+        seir_reward /= self._config['N_total']
```
Comment (r08922a27) [inline rl_reinforce.py:332]: 一樣 t 的部分看不懂，其他的 term 應該是基於某些 eval 的規則去訂的對吧？
麻煩之後說明一下了~

**Comment by r08922a27** [comment]
assert_all 我們應該要在加一行：
assert self.S + self.E + self.E_move + self.Q + self.Q_move + self.I + self.R == self._config['N_total']

File: config.json (hunk 1 above)
Comment (r08922196) [reply config.json:4]: Done

File: config.json (hunk 2 above)
Comment (r08922196) [reply config.json:8]: Done

**Comment by r08922196** [comment]
> assert_all 我們應該要在加一行：
> assert self.S + self.E + self.E_move + self.Q + self.Q_move + self.I + self.R == self._config['N_total']

Done

File: rl_reinforce.py (hunk 4 above)
Comment (r08922196) [reply rl_reinforce.py:327]: 調出來的，後期剩下來的 S 越多越重要。

File: rl_reinforce.py (hunk 5 above)
Comment (r08922196) [reply rl_reinforce.py:332]: 你是註解看不懂還是註解跟程式對不上？

前期出現 I Q E 要給予比較重的 penalty。

我這個不是針對 eval 去訂的，但是你可以從我整個 reward 的修改去推敲，我會更新在這個 PR 的說明。

File: rl_reinforce.py (hunk 3 above)
Comment (r08922196) [reply rl_reinforce.py:324]: t + 10 只是為了 smooth，

3 跟 2 是調出來的，原因我寫在註解上。

File: rl_reinforce.py (hunk 5 above)
Comment (r08922a27) [reply rl_reinforce.py:332]: 我看你的 code 是猜你的 t + 10 或是  t + 31 是為了讓這個 reward 可以考慮其他時期的狀況，我看不懂的地方是 t + const 怎麼影響到那個預期的前期或是後期，你可以舉個例子嗎？
譬如前期 t 是 5 的時候，設 t + 10 跟 reward 在前期重要的關係是什麼？ 因為後期譬如 t 假設可能 200 好了 但他還是 t + 10 啊？ 還是你是指量級不一樣 200 + 10 跟兩百差不多，5 跟 15 差很多？

File: rl_reinforce.py (hunk 4 above)
Comment (r08922a27) [reply rl_reinforce.py:327]: 取 log 有特別原因還是也是調出來的呢？

Comment (r08922196) [reply rl_reinforce.py:327]: 調出來的，你可以試試看拿掉，反正都是嚴格遞增

File: rl_reinforce.py (hunk 5 above)
Comment (r08922196) [reply rl_reinforce.py:332]: 1/1 = 1
1/(1+10) = 0.09

1/100 = 0.01
1/110 = 0.01

我想讓他前期平滑
